Add heuristics using stage spill metrics to skip apps

parthosa opened the pull request:
Fixes #477. This PR adds a generic `Additional Heuristics` module to skip recommending apps based on heuristics. It introduces an additional column `Skip By Heuristics` in qualification summary file. This logic will be applied only if the user tools is run with `--estimation_model xgboost` since it uses profiler output.

## Changes:

Added a specific heuristic to skip applications based on spill metrics:
1.  Check if profiler output is present, else skip the heuristics logic.
2. For each application:
   - Using `job_+_stage_level_aggregated_task_metrics.csv` Identify stages with spills greater than a threshold
   - Using `sql_to_stage_information.csv`, check if above spill stages have Execs other than the ones allowed (Join, Aggregate or Sort) 
   - If there are stages with significant spills and the spills is from Execs other than the ones allowed  <br/>   => Column `Skip By Heuristics` would be `True` for this application.
3. Finally, while calculating SpeedUp Category, if `Skip By Heuristics` is `True` for the application, set the category to `Not Recommended`.

## Things to discuss:
1. Spill Threshold is currently set to 10 GB (configurable in `qualification-conf.yaml`
2. We will read 3 csv per app. This should not be a bottleneck as Core tools would still take majority of time.
   - QualX reads ~10 csvs per app.
3. Corner case: 
   -  There is an application that is re-run and we test for both eventlogs together
   -  For TCO, we will group them by name, in this case the col `Skip By Heuristics` will be aggregated using `any()` function.
   - So, if any one of the run was skipped, the grouped application will also be skipped.
  

## Steps to Evaluate:
- Manually set `"Memory Bytes Spilled":50000000'` in the event `SparkListenerTaskEnd` for certain stages in any test event log.

tgravescs commented on user_tools/src/spark_rapids_tools/tools/additional_heuristics.py:
```diff
@@ -0,0 +1,120 @@
+        spill_threshold_bytes = self.props.get_value('spillBased', 'spillThresholdBytes')
+        stages_with_spills = job_stage_agg_metrics[
+            job_stage_agg_metrics['ID'].str.startswith('stage') &
+            (job_stage_agg_metrics['diskBytesSpilled_sum'] +
```
you are adding diskBytesSpilled to memoryBytesSpilled here, that is not correct. on a CPU event log those should not be added.  its a single spill and each of those represent sizes as if they were in memory vs compressed on disk.  
I would likely just stick to one, probably memory unless we have more data since the disk size compression could have more of an effect

parthosa replied:
Thanks @tgravescs. Checking the memory spill makes sense. I will update the logic.

tgravescs commented:
does this have any output saying which execs/stages/sqlid cause this to be skipped? 
Does this have output to explain why it should be skipped?

Did we generate any test eventlogs/queries that can be used to continue integration testing this or that could also be used by qualx to train on this scenario?

It seems like this should be done in Java, is there a followup to move it there?

parthosa commented:
> does this have any output saying which execs/stages/sqlid cause this to be skipped?
> Does this have output to explain why it should be skipped?

There is no output associated with it. A column `Skip By Heuristics Reason` can be added that mentions the details.
`StageId <stage_id> had <spill_size> spill`


> Did we generate any test eventlogs/queries that can be used to continue integration testing this or that could also be used by qualx to train on this scenario?

I have test event logs that I used to test this scenario. This can be added in the integration testing. I will include this as part of improving E2E tools testing #970

parthosa commented:
Introduced column `Skip By Heuristics Reason` column to state the reason:

There could be two potential reasons:
1. Spilling occurred -  _We should skip the app based on heuristics_
2. Profiler did not generate relevant output for the app -  _We should not skip the app based on heuristics (other reasons may still be applied)_

### Output Covering both cases:

File: `qualification_summary.csv` (formatted here and removed non-relevant columns for better readability):
```
|---|--------|-----------|------------|-----------------------|---------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------|
|   | Vendor | App Name  | App ID     | Estimated GPU Speedup | Speed Up Estimation Model | Skip by Heuristics | Skip by Heuristics Reason                                                                                                                         | Estimated GPU Speedup Category |
|---|--------|-----------|------------|-----------------------|---------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------|
| 0 | onprem | test-app0 | app-363454 |                  8.77 | XGBOOST                   | False              |                                                                                                                                                   | Large                          |
|---|--------|-----------|------------|-----------------------|---------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------|
| 1 | onprem | test-app1 | app-123456 |                  7.02 | XGBOOST                   | True               | Skipping due to spills in stages [39, 41, 40] exceeding 1000000000 bytes                                                                          | Not Recommended                |
|---|--------|-----------|------------|-----------------------|---------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------|
| 2 | onprem | test-app2 | app-654421 |                  1.00 | SPEEDUPS                  | False              | Cannot apply heuristics for qualification. Reason - FileNotFoundError:[Errno 2] No such file or directory: '/<path>/sql_to_stage_information.csv' | Not Recommended                |
|---|--------|-----------|------------|-----------------------|---------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------|
```


### Note to reviewers:
- If there are multiple apps with same app name, while generating the CSV for , they will be grouped.
   - In that case - `Skip By Heuristics Reason` column will be concatenated version of the individual reasons.

parthosa commented on user_tools/src/spark_rapids_pytools/rapids/qualification.py:
```diff
@@ -438,10 +439,25 @@ def __group_apps_by_name(self, all_apps) -> (pd.DataFrame, str):
+            """
+            Return a custom aggregation function or the input string.
+            """
+            agg_func_map = {
```
This is needed for concatenating `Skip By Heuristics Reason` column

parthosa commented:
## Changes
1. Create a new directory `intermediate_output` to store all intermediate output generated by user tools. We should avoid putting too much information in the `qualification_summary.csv`
2. Create a new file `heuristics_info.csv` in the above directory to store `[App ID,Skip by Heuristics,Reason]`


## Reasons
There could be two potential reasons:
1. Spilling occurred -  _We should skip the app based on heuristics_
3. Profiler did not generate relevant output for the app -  _We should not skip the app based on heuristics (other reasons may still be applied)_

## Output Covering both cases:

File:  `intermediate_output/heuristics_info.csv`
```
|--------------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| App ID                         | Skip by Heuristics | Reason                                                                                                                                            |
|--------------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| application_1686676198636_0003 | True               | Skipping due to spills in stages [39; 41; 40] exceeding 1000000000 bytes                                                                          |
|--------------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| application_1686676198636_0002 | False              |                                                                                                                                                   |
|--------------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| app-20231212214826-0000        | False              | Cannot apply heuristics for qualification. Reason - FileNotFoundError:[Errno 2] No such file or directory: '/<path>/sql_to_stage_information.csv' |
|--------------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| app-20240312004226-0000        | True               | Skipping due to spills in stages [60; 58] exceeding 1000000000 bytes                                                                              |
|--------------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| app-20240312023625-0000        | False              |                                                                                                                                                   |
|--------------------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|


```

amahussein reviewed (commented):
@mattahrens Are you ok with the PR as a temp work around until the heuristics are implemented in Scala module?

mattahrens commented:
Yes, I'm fine with it.  👍

amahussein reviewed (dismissed):
LGTME.
Thanks @parthosa

cindyyuanjiang commented on user_tools/src/spark_rapids_tools/tools/speedup_category.py:
```diff
@@ -75,12 +75,15 @@ def __process_category(self, all_apps: pd.DataFrame) -> pd.DataFrame:
-                if not entry.get('lowerBound') <= col_value <= entry.get('upperBound'):
+                # If the row is marked to be skipped by heuristics or the value is not within the range,
+                # set the category to default category (Not Recommended)
+                if (single_row.get(heuristics_col_name) is True or
```
minor nit: why do we need `is True`?

parthosa replied:
I had two concerns from `if single_row.get(heuristics_col_name) or` approach:
1. I wanted to keep this explicit for more clarity on what the field is supposed to return
2. Since it is not typed, `single_row.get(heuristics_col_name)` could return any value and the condition would allow it to pass

cindyyuanjiang reviewed (dismissed):
Thanks @parthosa! Just a minor nit.
I am also wondering in the output file: is `1000000000 bytes` or `10 GB` more clear?

parthosa commented:
> I am also wondering in the output file: is 1000000000 bytes or 10 GB more clear?

Thanks @cindyyuanjiang. I think 10 GB would be more clear. Added a function to convert bytes to human readable format. We have the following reason now:
```
App ID,Skip by Heuristics,Reason
app-20240312004226-0000,True,Skipping due to spills in stages [60; 58] exceeding 10.00 GB
```

cindyyuanjiang reviewed (approved):
Thanks @parthosa! LGTM.
